Release Script: Support multiple plugins

sdixon194 opened the pull request:
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->


## Proposed changes:
<!--- Explain what functional changes your PR includes -->
This updates the release-plugin.sh script to add support for multiple plugins, and will automatically add the mu-wpcom-plugin if we're releasing jetpack.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
N/a
## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
no
## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

Hard to test so might be best to test next cycle. 

It should work by running `tools/release-plugin.sh jetpack 12.0-a.0` etc. Support for multiple plugins like `tools/release-plugin.sh jetpack 12.0-a.0 backup 1.5 search 1.5.0` etc. If no release version is specified, it will try to automatically determine the next versions.

sdixon194 committed:
Add the mu-wpcom-plugin to the array

sdixon194 committed:
Prompt for a version number

sdixon194 committed:
Fix checking for plugins

sdixon194 committed:
Confirm everything before starting the process

sdixon194 committed:
Start getting the prefixes

sdixon194 committed:
Check for existing release branches

sdixon194 committed:
Testing commits

sdixon194 committed:
Fix how we handle the beta

sdixon194 committed:
Start the release branch

sdixon194 committed:
More release branch

sdixon194 committed:
Fixup the release branch

sdixon194 committed:
Uncomment the live pushes

sdixon194 committed:
Update the usage

sdixon194 committed:
remove uneeded includes

anomiex commented on tools/release-plugin.sh:
```diff
@@ -67,30 +56,67 @@ while getopts "v:abh" opt; do
+	if [[ "$1" =~ ^[a-zA-Z\-]+$ ]]; then
+		PLUGIN=$1
+	else
+		die "Script must be passed in [project] <version> format. Got $1"
```
The usual convention is to use `[]` to delimit optional arguments, so I'd do like
```suggestion
		die "Script must be passed in <project> [<version>] format. Got $1"
```
or maybe
```suggestion
		die "Script must be passed in <project> [version] format. Got $1"
```

sdixon194 replied:
Oh! I had them backwards, thank you.

anomiex commented on tools/release-plugin.sh:
```diff
@@ -67,30 +56,67 @@ while getopts "v:abh" opt; do
+	fi
+	
+	if [[ "$2" =~ ^[0-9]+(\.[0-9]+)+(-.*)?$ ]]; then
+		PROJECTS["$PLUGIN"]+=$2
```
Is there a reason for `+=` here instead of just `=`?

sdixon194 replied:
Nope, just mixed up my syntax! 4c09f2c2184559da8bc22f761e8c2218dad1e707

anomiex commented on tools/release-plugin.sh:
```diff
@@ -67,30 +56,67 @@ while getopts "v:abh" opt; do
-	yellow "Project found: $SLUG"
-	PROJECT="$SLUG"
+# If we're releasing Jetpack, we're also releasing mu-wpcom-plugin.
+if [[ "${!PROJECTS[*]}" =~ "jetpack" ]]; then
```
I don't like this, it's liable to false-positive if any plugin's name happens to contain the string "jetpack". Why not test directly? (and while we're at it, avoid overwriting a passed version for mu-wpcom-plugin)
```suggestion
if [[ -n "${PROJECTS[jetpack]+ok}" && -z "${PROJECTS[mu-wpcom-plugin]+ok}" ]]; then
```

anomiex commented on tools/release-plugin.sh:
```diff
@@ -67,30 +56,67 @@ while getopts "v:abh" opt; do
+for SLUG in "${!PROJECTS[@]}"; do
+	if [[ -z "${PROJECTS[$SLUG]}" ]]; then
+		cd "$BASE"
+		if [[ -x "projects/$SLUG/vendor/bin/changelogger" ]]; then
```
Note this test could fail if no one has done `jetpack install --all` (or `jetpack install $SLUG` or `composer install` in the plugin directory).

Most of the scripts do something like this, IIRC
```
CL="$BASE/projects/packages/changelogger/bin/changelogger"
if ! "$CL" &>/dev/null; then
    (cd "$BASE/projects/packages/changelogger" && composer update --quiet)
    if ! "$CL" &>/dev/null; then
        die "Changelogger is not runnable via $CL"
    fi
fi
```
although `changelogger-release.sh` does `pnpm jetpack install --all` so it can do like this and `tools/changelogger-validate-all.sh` tests for `vendor/bin/changelogger` but falls back to `$BASE/projects/packages/changelogger/bin/changelogger`. Someday I should probably clean that up so everything does the same and without having to copy-paste a big block of code like the above.

anomiex replied:
> Someday I should probably clean that up so everything does the same and without having to copy-paste a big block of code like the above.

https://github.com/Automattic/jetpack/pull/29371

sdixon194 replied:
I think I updated it correctly in 645c5fe10baaffca1dad2c82f7944c621f95f720 

Do you think it's worth having the release command throw a `pnpm jetpack install --all` first to be extra super safe now?

anomiex replied:
We could, but I'd avoid it if we can since it does take a while to run.

anomiex commented on tools/release-plugin.sh:
```diff
@@ -67,30 +56,67 @@ while getopts "v:abh" opt; do
 	fi
 	local CUR_VERSION
-	CUR_VERSION=$("$BASE/tools/plugin-version.sh" "$PROJECT")
+	CUR_VERSION=$("$BASE/tools/plugin-version.sh" "$2")
```
I'd probably go ahead and make the project `$1` and the version `$2`, but it doesn't really matter.

You might also want to include the project in the messages on lines 115 and 122, for context. Or I suppose you could move line 132 to before the call on line 129 to supply the context.

sdixon194 replied:
be1f09f1dda3b8f65d6a6130e2bfc2405cc6858b

anomiex commented on tools/release-plugin.sh:
```diff
@@ -99,41 +125,43 @@ function check_ver {
-else
-	die "No release branch prefix found for $PROJECT, aborting."
+# If we're releasing Jetpack, we need to set that prefix first.
+if [[ "${!PROJECTS[*]}" =~ "plugins/jetpack" ]]; then
```
Same comment as line 83.

anomiex commented on tools/release-plugin.sh:
```diff
@@ -99,41 +125,43 @@ function check_ver {
-	RELEASE_BRANCH="$PREFIX/branch-${VERSION%%-*}"
-else
-	die "No release branch prefix found for $PROJECT, aborting."
+# If we're releasing Jetpack, we need to set that prefix first.
```
Why?

sdixon194 replied:
This was based on my misunderstanding of the release script where I thought if the script assumed we were releasing backup or search by hitting that plugin first, it would use a different branch prefix, but obviously that doesn't happen.

sdixon194 replied:
I ended up adding this back - I realized when testing something else:

- On line 146 we create the associative array of the prefixes in the form of prefix => version
- if we're releasing, say, jetpack and backup, the loop finds backup first and assigns it like `jetpack => <backup's version>`
- We need to check for existing release branches later in line 158 using the values from the prefixes array. If we checked anything but jetpack-the-plugin first earlier, it won't find the existing branch since it'll look for something like `jetpack/branch-1.6` (or whatever backup's version is) instead of `jetpack/branch-12.0` etc. 

Maybe there's another way to work around this?

anomiex replied:
When preparing my update for the changes in #29707, I wound up letting `tools/create-release-branch.sh` handle deleting any existing release branch instead of doing it here. I don't know whether we'll wind up doing like "monthly/branch-2023-03-28" now or still like "monthly/branch-12.0" or what, but it's not so clearly depending on a version number anymore.

anomiex commented on tools/release-plugin.sh:
```diff
@@ -99,41 +125,43 @@ function check_ver {
-fi
+for SLUG in "${!PROJECTS[@]}"; do
+	PREFIX=$(jq -r '.extra["release-branch-prefix"] // ""' "$BASE"/projects/"$SLUG"/composer.json)
+	if [[ -n "$PREFIX" ]] && [[ ! "${PREFIXES[*]}" =~ $PREFIX ]]; then
```
Same as line 83 again. Plus that will avoid trouble if `$PREFIX` ever happens to contain any regex special characters for some reason.

Also, do like `[[ test1 && test2 ]]` instead of `[[ test1 ]] && [[ test2 ]]`.

sdixon194 replied:
Updated in fba64bd72e3ba38503bf17fa397e3f101bb053a4

anomiex commented on tools/release-plugin.sh:
```diff
@@ -158,21 +186,41 @@ if ! git push -u origin HEAD; then
+for PLUGIN in "${!PROJECTS[@]}"; do
+	yellow "Updating the changelog files for $PLUGIN."
+	
+	if [[ ${ARGS} ]]; then
+		unset ARGS
+	fi
```
This is probably unnecessary. Line 198 should overwrite it anyway.

sdixon194 replied:
d783b184a76f31587333871ae485acc436db333f

anomiex commented on tools/release-plugin.sh:
```diff
@@ -158,21 +186,41 @@ if ! git push -u origin HEAD; then
-if [[ "$PROJECT" == "projects/jetpack" && "$ALPHABETA" == "-b" ]]; then
-	yellow "Releasing a beta, amending the readme.txt"
+# If we're releasing Jetpack and it's a beta, amend the readme.txt
+if [[ -v PROJECTS["plugins/jetpack"] && "${PROJECTS[plugins/jetpack]}" =~ '-beta' ]]; then
```
Ooh, `-v` is even better than what I suggested back on line 83.

Probably better to test `== *-beta` instead of `=~ '-beta'` though, no need for regex when globbing will do.

sdixon194 replied:
> Ooh, -v is even better than what I suggested back on line 83.

Ah yes, this is probably a chatGPT suggestion :D 

a7874b27a7425c07c5aa40c52cea3d47fb9adefb

anomiex commented on tools/release-plugin.sh:
```diff
@@ -201,8 +249,14 @@ if ! gh run watch "${BUILDID[0]}" --exit-status; then
-tools/create-release-branch.sh "$PROJECT" "$VERSION"
+yellow "Build is complete."
+# Run tools/create-release-branch.sh to create a release branch for each project.
+for PREFIX in "${!PREFIXES[@]}"; do
+	git checkout prerelease
+	VERSION="${PREFIXES[$PREFIX]}"
+	PROJECT="$PREFIX"
+	yellow "Creating release branch for $PROJECT $VERSION"
+	tools/create-release-branch.sh "$PROJECT" "$VERSION"
+done
```
is this going to do the right thing, or will it wind up confused because the prerelease branches each have extraneous plugins also in the release state? Do we really need to handle the case of releasing more than one different prefix at once?

sdixon194 replied:
Hmm, how would we do this normally? Occasionally a team will find out we are doing a release and ask us to release it alongside Jetpack, but I haven't tested that flow since we used the prerelease branch.

Can they all just be done from the same release branch? Something like Protect can just be released from `jetpack/branch-12.0`, etc?

anomiex replied:
Looks like CI will be ok after all, now that I look closer for a release branch it only checks the relevant plugins.

sdixon194 committed:
Update tools/release-plugin.sh

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

sdixon194 committed:
Update tools/release-plugin.sh

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

sdixon194 committed:
Updated adding element to array with correct syntax

sdixon194 committed:
Merge remote-tracking branch 'origin/trunk' into update/release-script-mu-plugin

sdixon194 committed:
Update how changelogger is invoked

sdixon194 committed:
Swap project and version number when calling check_ver

sdixon194 committed:
Remove need to check for Jetpack first

sdixon194 committed:
Remove uneeded array unset

sdixon194 committed:
Fix handling how we check for versions ending in -beta

sdixon194 committed:
Check for associative array value using -v

sdixon194 committed:
[not verified] Revert "Remove need to check for Jetpack first"

This reverts commit 26e825acea4e7f4edfac2cf3469ad390f0459c9b.

sdixon194 committed:
Check if we're releasing jetpack during the prefix

anomiex committed:
Update for PR 29707

* Handle multiple configured release branch prefixes, preferring ones
  that cover multiple plugins when appropriate.
* Leave deleting existing release branches to
  `tools/create-release-branch.sh` as predicting the suffix is harder
  now.

anomiex committed:
Sensible defaults in create-release-branch

When prompting for versions, pre-fill with the current version since
that's most commonly what's going to be used. Also don't bother
confirming when we're using the current version, that's the usual case.

kraftbj committed:
Merge remote-tracking branch 'origin/trunk' into update/release-script-mu-plugin

kraftbj commented:
@anomiex Would you mind giving this a review after the latest changes from Steve. If there's anything outstanding, would like to see if Steve can address or provide some notes tomorrow before his leave.

anomiex commented:
Yesterday I was working on an update to this PR to take into account the changes in #29707, other than that I think we're pretty good here. I held off on pushing anything just because GH doesn't handle PRs depending on other PRs all that well. 

My impression was that Steve already effectively handed this off to the team to finish last week, but I'd be happy to consider any additional notes he has on it.

sdixon194 commented:
> My impression was that Steve already effectively handed this off to the team to finish last week, but I'd be happy to consider any additional notes he has on it.

Thanks for working on this Brad! I don't have any additional notes at the moment. Sounds like we're waiting on https://github.com/Automattic/jetpack/pull/29707, which is waiting until after the release maybe. It looks like all my previous comments were addressed. Sounds like you have a commit ready to go to update this one, so I won't work on anything local.

anomiex committed:
Merge remote-tracking branch 'origin/trunk' into update/release-script-mu-plugin

anomiex committed:
Merge remote-tracking branch 'origin/trunk' into update/release-script-mu-plugin

samiff commented on tools/release-plugin.sh:
```diff
@@ -67,30 +57,68 @@ while getopts "v:abh" opt; do
+	SLUG="${SLUG%/}" # Sanitize
+	SLUG="plugins/$SLUG"
+	if [[ ! -e "$BASE/projects/$SLUG/composer.json" ]]; then
+		die "$SLUG isn't a valid project!"
```
@anomiex Maybe I'm a bit confused, but running with `tools/release-plugin.sh weekly` seems to exit here.

anomiex replied:
You are a bit confused. You're supposed to do like `tools/release-plugin.sh jetpack mu-wpcom-plugin` and the script will decide that "weekly" is the prefix to use.

Or, for this specific case, `tools/release-plugin.sh jetpack` will work because it's coded to assume `mu-wpcom-plugin` always goes with Jetpack.

anomiex committed:
Fix cwd

samiff commented:
@zinigor @bindlegirl Let's give this a try for the next release that goes out.
